internal/controller/supervisorconfig/githubupstreamwatcher/github_upstream_watcher.go

```diff
@@ -93,52 +115,328 @@ func (c *gitHubWatcherController) Sync(ctx controllerlib.Context) error {
+	}, policy
 }
 
-func (c *gitHubWatcherController) validateUpstream(ctx controllerlib.Context, upstream *v1alpha1.GitHubIdentityProvider) (*upstreamgithub.ProviderConfig, error) {
-	result := upstreamgithub.ProviderConfig{
-		Name: upstream.Name,
+func (c *gitHubWatcherController) validateUpstreamAndUpdateConditions(ctx controllerlib.Context, upstream *v1alpha1.GitHubIdentityProvider) (
+	*upstreamgithub.ProviderConfig, // If validated, returns the config
+	error, // This error will only refer to programmatic errors such as inability to perform a Dial or dereference a pointer, not configuration errors
+) {
+	conditions := make([]*metav1.Condition, 0)
+	applicationErrors := make([]error, 0)
+
+	clientSecretCondition, clientID, clientSecret, clientSecretErr := c.validateClientSecret(upstream.Spec.Client.SecretName)
+	conditions = append(conditions, clientSecretCondition)
+	if clientSecretErr != nil {
+		applicationErrors = append(applicationErrors, clientSecretErr)
+	}
+
+	// Should there be some sort of catch-all condition to capture this?
+	// This does not actually prevent a GitHub IDP from being added to the cache.
+	groupNameAttribute, usernameAttribute, userAndGroupErr := validateUserAndGroupAttributes(upstream)
+	// Should not happen due to CRD defaulting
+	if userAndGroupErr != nil {
+		applicationErrors = append(applicationErrors, userAndGroupErr)
+	}
+
+	organizationPolicyCondition, policy := validateOrganizationsPolicy(upstream)
+	conditions = append(conditions, organizationPolicyCondition)
+
+	hostAndTLSConditions, host, httpClient, validateHostAndTLSErr := c.validateHostAndTLSConfiguration(ctx.Context, upstream.Spec.GitHubAPI)
+	conditions = append(conditions, hostAndTLSConditions...)
+	if validateHostAndTLSErr != nil {
+		applicationErrors = append(applicationErrors, validateHostAndTLSErr)
+	}
+
+	if len(conditions) != 5 {
+		applicationErrors = append(applicationErrors, fmt.Errorf("expected %d conditions but found %d conditions", 5, len(conditions)))
+		return nil, errors.NewAggregate(applicationErrors)
 	}
 
-	// TODO: once we firm up the proposal doc & merge, then firm up the CRD & merge, we can fill out these validations.
 	// The critical pattern to maintain is that every run of the sync loop will populate the exact number of the exact
 	// same set of conditions.  Conditions depending on other conditions should get Status:  metav1.ConditionUnknown, or
 	// Status:  metav1.ConditionFalse, never be omitted.
-	conditions := []*metav1.Condition{
-		// we may opt to split this up into smaller validation functions.
-		// Each function should be responsible for validating one logical unit and setting one condition.
-		// c.validateGitHubAPI(),
-		// c.validateClaims(),
-		// c.validateAllowedAuthentication(),
-		// c.validateClient(),
+
+	hasErrorCondition := conditionsutil.HadErrorCondition(conditions)
+	updateStatusErr := c.updateStatus(ctx.Context, upstream, conditions)
```

Comment:
The order here is a little strange:
```
- hasErrorCondition (check)
- updateStatusError (possible new error returned after the check)
- if special updateStatusError != nil (special check)
- if hasErrorCondition (back to original check)
```

Can we swap these lines around or is there a subtle function in here that depends on the ordering?